### PR TITLE
feat(prediction): prediction 2D 

### PR DESCRIPTION
**Pull request by jobo322**
add statistic when it is available

### CODE REVIEW

File: src/prediction/data/carbonDB-withoutStatistic.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+export default [
```
Comment (lpatiny) [inline src/prediction/data/carbonDB-withoutStatistic.js:1]: This file should move to the test folder. We should in the test send the database.

File: src/prediction/data/carbonDB-withStatistic.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+export default [
```
Comment (lpatiny) [inline src/prediction/data/carbonDB-withStatistic.js:1]: This file should move in the test folder

File: src/prediction/predictAll.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import { predictCarbon } from './predictCarbon';
+import { predictHMBC } from './predictHMBC';
+import { predictHSQC } from './predictHSQC';
+import { predictProton } from './predictProton';
+
+/**
+ * Generate the correlation bidimensional nmr peaks based on the number of bonds between a pair of atoms
+ * @param {Molecule} molecule - openchemlib molecule instance
+ * @param {object} [options = {}] - options
+ * @param {number} [options.minLength = 1] - minimum number of bonds to take into account.
```
Comment (lpatiny) [inline src/prediction/predictAll.js:11]: minLength and maxLength should be removed

File: src/prediction/predictAll.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ * @param {number} [options.maxLength = 1] - maximum number of bonds to take into account.
+ * @param {object} [options.predictOptions={}] - options for predictors. The object should has properties
+ * with name equal to the value of options.from and options.to.
+ * @param {object} [options.predictions={}] - mono dimensional signal to generate the bidimenionals cross peaks.
+ * The object should has properties with name equal to "H" and "C".
+ * the value of each prediction should be similar that predictProton and predictCarbon.
+ * @param {object} [options.predictor={H: predictProton, C: predictCarbon}] - mono dimensional predictors. The object should has properties with name
+ * equal to "H" and "C".
+ * @param {number} [options.joinDistance=0.05] - signals with euclidean distance in ppm less than this it will
```
Comment (lpatiny) [inline src/prediction/predictAll.js:20]: We need to have preferences per nucleus. An object with {H: 0.05, C: 0.5}  ?

File: src/prediction/predictProton.js [hunk 5]
```diff
@@ -14,7 +14,7 @@ import { signalsToRanges } from '../signals/signalsToRanges.js';
  * @param {Molecule} molecule - OCL Molecule instance.
  * @param {object} [options={}]
  * @param {function} [options.cache] A callback receiving a molfile and the result
- * @return {Promise<Array>}
+ * @returns {Promise<Array>} - object with molfile, diaIDs, signals, joined signals by diaIDs and ranges.
```
Comment (lpatiny) [inline src/prediction/predictProton.js:17]: Why Array ?

File: src/prediction/utils/predict2D.js [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+  // clusterize signals by distance
+  const signals = Object.values(group);
+  const zones = joinSignals(signals, { joinDistance });
+
+  return {
+    molfile: molecule.toMolfile(),
+    diaIDs: spectra.x.diaIDs,
+    signals,
+    zones,
+  };
```
Comment (lpatiny) [inline src/prediction/utils/predict2D.js:118]: For 1D we have:
```
    joinedSignals,
    signals,
    ranges: signalsToRanges(joinedSignals)
```

Here will also need to join the signals if they are between the same diastereotopic IDs.

**Comment by lpatiny** [comment]
Nice work.

Please could you also create (in another pull request) the generation of a 2D spectrum based on the signals ? I guess we need to create a matrix + min,max in X and Y.

Like that NMRium can regenerate the contour plot

File: src/prediction/predictAll.js (hunk 4 above)
Comment (jobo322) [reply src/prediction/predictAll.js:20]: The joinDistance is the distance in the bidimensional space. could we calculate the joinDistance as sqrt(H*H + C*C) and H and C the preferences peer nucleus?

File: src/prediction/predictProton.js (hunk 5 above)
Comment (jobo322) [reply src/prediction/predictProton.js:17]: It is a copy paste error from predictionProton, it is wrong. It is a Promise<object> right?

File: src/prediction/utils/predict2D.js (hunk 6 above)
Comment (jobo322) [reply src/prediction/utils/predict2D.js:118]: actually, signals is all ready joined by diasterotopic IDs. do you want both?

Comment (jobo322) [reply src/prediction/utils/predict2D.js:122]: done

File: src/prediction/predictProton.js (hunk 5 above)
Comment (jobo322) [reply src/prediction/predictProton.js]: done

**Pull request merged**